File: ecmascript-testcases/test-dev-internal-property-basics.js
/*
 *  Since Duktape 0.12.0, user code can also employ internal properties.
 *  This testcase exercises a few basics of that.
 */

/*---
{
    "custom": true
}
---*/

/*===
for-in: foo
for-in:  <255>quux
JSON.stringify: {"foo":1," <195><191>quux":4}
JX encoded: {foo:1," \xffquux":4}
Object.keys: foo, <255>quux
Object.getOwnPropertyNames: foo,bar, <255>quux
direct access: 3
===*/

function sanitizedPrint(x) {
    var buf, i, n;
    var tmp = '';

    buf = Duktape.Buffer(x);
    for (i = 0, n = buf.length; i < n; i++) {
        if (buf[i] < 0x20 || buf[i] > 0x7e || buf[i] == '<' || buf[i] == '>') {
            tmp += '<' + String(buf[i]) + '>';
        } else {
            tmp += String.fromCharCode(buf[i]);
        }
    }

    print(tmp);
}

function test() {
    var internalKey;
    var invalidUtf8Key;
    var obj;

    // 'foo' is enumerable, writable, configurable
    var obj = { foo: 1 };

    // 'bar' is not enumerable
    Object.defineProperty(obj, 'bar', {
        value: 2, writable: true, enumerable: false, configurable: true
    });

    // Internal key \xFF\xFFabc is in principle enumerable, but because
    // internal keys have special behavior, it is never enumerated.
    internalKey = Duktape.dec('hex', 'ffff616263');  // \xFF\xFFabc
    obj[internalKey] = 3;

    // The key \x20\xFFquux is invalid UTF-8 but not an internal string,
    // so the property will be enumerable.  Behavior for invalid UTF-8
    // varies.  For example, JSON.stringify() will encounter the invalid
    // UTF-8 initial byte \xFF and serialize it like it had encountered
    // the codepoint U+00FF (writing out the bytes c3 bf).
    invalidUtf8Key = Duktape.dec('hex', '20ff71757578');  // \x20\xFFquux
    obj[invalidUtf8Key] = 4;

    // For-in only lists enumerable keys
    for (k in obj) {
        sanitizedPrint('for-in: ' + k);
    }

    // JSON.stringify() also only lists enumerable keys
    sanitizedPrint('JSON.stringify: ' + JSON.stringify(obj));

    // Even the Duktape-specific JX format doesn't list internal keys
    sanitizedPrint('JX encoded: ' + Duktape.enc('jx', obj));

    // Object.keys() only returns enumerable keys (but not internal keys,
    // even if they're technically enumerable)
    sanitizedPrint('Object.keys: ' + Object.keys(obj));

    // Object.getOwnPropertyNames() returns also non-enumerable keys,
    // but not internal keys
    sanitizedPrint('Object.getOwnPropertyNames: ' + Object.getOwnPropertyNames(obj));

    // Direct property access to the internal property still works.
    sanitizedPrint('direct access: ' + obj[internalKey]);
}

try {
    test();
} catch (e) {
    print(e);
}

File: api-testcases/test-dev-internal-property-basics.c
/*
 *  Example of using internal properties from C code.
 */

/*===
*** test_1 (duk_safe_call)
{foo:1," \xffbar":3}
final top: 1
==> rc=0, result='undefined'
===*/

static duk_ret_t test_1(duk_context *ctx) {
	duk_eval_string(ctx, "(function (x) { print(Duktape.enc('jx', x)); })");

	duk_push_object(ctx);

	/* Ordinary property */
	duk_push_int(ctx, 1);
	duk_put_prop_string(ctx, -2, "foo");  /* obj.foo = 1 */

	/* Internal property \xFF\xFFabc, technically enumerable (based on
	 * property attributes) but because of internal property special
	 * behavior, does not enumerate.
	 */

	duk_push_int(ctx, 2);
	duk_put_prop_string(ctx, -2, "\xff\xff" "abc");  /* obj[\xff\xffabc] = 2, internal property */

	/* Another property with invalid UTF-8 data but doesn't begin with
	 * \xFF => gets enumerated and JX prints out an approximate key.
	 */
	duk_push_int(ctx, 3);
	duk_put_prop_string(ctx, -2, " \xff" "bar");  /* obj[ \xffbar] = 3, invalid utf-8 but not an internal property */
	duk_call(ctx, 1);

	printf("final top: %ld\n", (long) duk_get_top(ctx));
	return 0;
}

void test(duk_context *ctx) {
	TEST_SAFE_CALL(test_1);
}
